app.js

```diff
@@ -68,6 +68,222 @@ function nonce( length ) {
+  if ( !validation.client_id ) {
+    console.error( "Invalid Password" );
+    return null;
+  }
+
+  const missingScopes = requiredScopes.filter( scope => !validation.scopes.includes( scope ) );
+
+  if ( missingScopes.length ) {
+    console.error( "Missing required scopes: ", missingScopes.join(", ") );
+    return null;
+  }
+
+  return validation.client_id;
+}
+
+/**
+ * 
+ * @param { string } channel 
+ * @param { string } clientId 
+ * @param { string } password 
+ * @returns { string | null }
+ */
+async function fetchChannelIdAsync( channel, clientId, password ) {
+  let err;
+  let userInfo = await fetch( "https://api.twitch.tv/helix/users?login=" + channel, {
+    headers: {
+      "Client-ID": clientId,
+      "Authorization": `Bearer ${password}`
+    }
+  }).then( r => r.json() )
+  .catch( e => (err = e, null));
+
+  if ( err || userInfo === null ) {
+    console.error( "Error fetching user info: ", err );
+    return null;
+  }
+
+  return userInfo.data[ 0 ].id;
+}
+
+/**
+ * 
+ * @param { string } type 
+ * @param { string } version 
+ * @param { string } clientId 
+ * @param { string } password 
+ * @param { string } channelId 
+ * @param { string } sessionId 
+ * @returns { boolean } was the subscription successful
+ */
+async function subscribeToEventAsync( type, version, clientId, password, channelId, sessionId ) {
+  let err;
+  await fetch( "https://api.twitch.tv/helix/eventsub/subscriptions", {
+    method: "POST",
+    headers: {
+      "Client-ID": clientId,
+      "Authorization": `Bearer ${password}`,
+      "Content-Type": "application/json"
+    },
+    body: JSON.stringify( {
+      type,
+      version,
+      condition: {
+        broadcaster_user_id: channelId,
+      },
+      transport: {
+        method: "websocket",
+        session_id: sessionId
+      }
+    } )
+  })
+  .catch( e => (err = e, console.error( "Error subscribing to event: ", type, e )));
+
+
+  return !err;
+}
+
+async function eventSubConnectAsync( channel, password, clientId = null, channelId = null, connectionName = null, sessionId = null) {
+  /** @type { [string, string][] } */
+  const subscribtions = [
+    [ "channel.channel_points_automatic_reward_redemption.add", "1" ],
+    [ "channel.channel_points_custom_reward_redemption.add", "1" ],
+  ];
+
+	password = password.replace( "oauth:", "" );
+
+  if ( !clientId ) {
+    clientId = await fetchClientIdIfValidAsync( password );
+    if ( clientId === null ) {
+      return;
+    }
+  }
+
+  if ( !channelId ) {
+    channelId = await fetchChannelIdAsync( channel, clientId, password );
+    if ( channelId === null ) {
+      return;
+    }
+  }
+
+  const keepAliveSeconds = 30;
+  const heartbeatInterval = 1000 * 60; //ms between PING's
+  if (!connectionName) {
+    connectionName = "wss://eventsub.wss.twitch.tv/ws";
+    if ( keepAliveSeconds !== 30 ) {
+      connectionName += "?keepalive_timeout_seconds=" + keepAliveSeconds;
+    }
+  }
+
+  const ws = typeof window !== "undefined"
```

Comment:
I think it might be good to be able to have this websocket available outside of the function or to return it after the connect completes to be able to call it and disconnect or do other things with it. What do you think?

Reply:
are we planning on using it outside point rewards?

https://dev.twitch.tv/docs/eventsub/handling-websocket-events/#revocation-message
>IMPORTANT By default, you have 10 seconds from the time you receive the Welcome message to subscribe to an event, unless otherwise specified when connecting. If you don’t subscribe within this timeframe, the server closes the connection.

I'm not sure if that applies to first subscription or the entire lifecycle

i think it would be reasonable to return the obscured `onDisconnect` function